draft-ietf-lwig-minimal-esp.xml

```diff
@@ -104,79 +104,78 @@ The primary purpose of Minimal ESP is to remain interoperable with other nodes i
+The use of a predictive function to generate the SPI might be preferred over the generation and handling of random values.
+An implementation of such predictable function could use the combination of a fixed value and the memory address of the SAD structure.
+For every incoming packet, the node will be able to point to the SAD structure directly from the SPI value.
+This avoids having a separate and additional binding and lookup function for the SPI to its SAD entry for every incoming packet.
 </t>
 
 <!-- privacy / security implication for non random SPIs -->
 
 <section title="Considerations over SPI generation">
 
-<t>SPI that are not randomly generated over 32 bits may lead to privacy and security concerns. 
-As a result, the use of alternative designs requires careful security and privacy reviews. 
+<t>SPIs that are not randomly generated over 32 bits may have privacy and security concerns.
+As a result, the use of alternative designs requires careful security and privacy reviews.
 This section provides some considerations upon the adoption of alternative designs. </t>
 
-<t>Note that SPI value is used only for inbound traffic, as such the SPI negotiated with IKEv2 <xref target="RFC7296"/> or <xref target="RFC7815"/> by a peer, is the value used by the remote peer when it sends traffic. 
-As SPI is only used for inbound traffic by the peer, this allows each peer to manage the set of SPIs used for its inbound traffic. 
-Similarly, the privacy concerns associated with the generation of non random SPI is also limited to the incoming traffic.</t>
+<t>The SPI value is only looked up for inbound traffic.
+The SPI negotiated with IKEv2 <xref target="RFC7296"/> or Minimal IKEv2 <xref target="RFC7815"/> by a peer is the value used by the remote peer when it sends traffic.
+As the SPI is only used for inbound traffic by the peer, this allows each peer to manage the set of SPIs used for its inbound traffic.
+Similarly, the privacy concerns associated with the generation of non-random SPIs is also limited to the incoming traffic.</t>
 
-<t>When alternate designs are considered, it is likely that the number of possible SPIs will be limited. 
-This limit should both consider the number of inbound SAs - possibly per IP addresses - as well as the ability for the node to rekey. 
-SPI can typically be used to implement a key update with the SPI indicating the key is being used. 
-For example, a SPI might be encoded with the Security Association Database (SAD) entry on a subset of bytes (for example 3 bytes), while the remaining byte indicates the rekey index. 
+<t>Alternate designs that take less resources than fully random SPI's are likely using a limited list of possible SPIs.
+This limit should take into account the number of inbound SAs - possibly per IP addresses - as well as the requirement for rekeying which would briefly require 2 inbound SPIs to co-exist as the new SA is setup before the old SA is torn down.
```

Comment:
I don't think whether or not a packet is sent is relevant? During the rekey you have multiple SAs and multiple SPIs. If implmenters are making decisions on the number of hardcoded SPIs to support for say 4 simultaneous IPsec tunnels, I think a clear warning you need 8 SPIs is justified.
 
Note I also do not really understand your modifier "possibly per IP address". Are you referring to failover links or HA?

Reply:
But if some other mechanism than IKE is used, why would that mechanism not provide both the SPI numbers and the policy/key information et all? I don't think your device would receive a crypto key and a packet policy but then be told to generate its own SPI for it ?

I thought you meant to say something like "I have 4 slots for SPIs, and I will use 257,258,259 and 260 and for rekeyed 513,514,515 and 516"  (and then flip between 257 and 513 every rekey), so in that way the SPI "indexes" if an SA is rekeyed or not.

If you are really talking about an IKE case and a non-IKE case, you should mention more clearly, but expect to provide  answers to AEADs and manual keying. I recommend you navigate away from this, and put all the responsibility for manual keying, including SPI generation, to a "secure IKE alternative provding full SA information to the constrained device"

Reply:
We do consider the case where IKE has to manage a limited number of slots as well as the case where the mote is hard coded. In the latter case, it might be entirely provisioned, and you need to provision the other part manually to exchange with the mote.

I do not see  the issue with AEAD, are you referring to the salt ?